src/dmd/root/file.d

```diff
@@ -22,6 +22,320 @@ import dmd.root.filename;
 import dmd.root.rmem;
 import dmd.root.string;
 
+/**
+Runs a non-pure function or delegate as pure code. Use with caution.
+Returns: whatever `fun` returns.
+*/
+private auto fakePure(F)(scope F fun) pure
+{
+    return (cast(mixin(F.stringof ~ " pure")) fun)();
+}
+
+/**
+Encapsulated management of a memory-mapped file.
+
+The `Datum` type encodes the mapping mode: Use `ubyte` for read/write mapping
+and `const ubyte` for read-only mapping.
+*/
+struct FileMapping(Datum)
+{
+    static assert(Datum.sizeof == 1);
+
+    version(Posix) enum invalidHandle = -1;
+    else version(Windows) enum invalidHandle = INVALID_HANDLE_VALUE;
+
+    /// Handle of underlying file
+    private auto handle = invalidHandle;
+    /// File mapping object needed on Windows
+    version(Windows) private HANDLE fileMappingObject = invalidHandle;
+    /// Memory-mapped array
+    private Datum[] data;
+    /// Name of underlying file, zero-terminated
+    private const(char)* name;
+
+    import dmd.backend.errors : error;
+    import dmd.errors : fatal;
+
+    invariant
+    {
+        if (handle != invalidHandle)
+        {
+            // Only named mappings supported for now.
+            assert(name !is null);
+            return;
+        }
+        version(Windows) assert(fileMappingObject == invalidHandle);
+        assert(data.ptr is null);
+        assert(name is null);
+    }
+
+    /**
+    Open `filename` and map in memory. If `Datum` is `const`, opens for
+    read-only and maps the content in memory; no error is issued if the file
+    does not exist. This makes it easy to treat a non-existing file as empty.
+
+    If `Datum` is mutable, opens for read/write (creates file if it does not
+    exist) and fails fatally on any error.
+
+    Due to quirks in `mmap`, if the file is empty, `handle` is valid but `data`
+    is `null`.
+    */
+    this(const char* filename)
```

Comment:
Should there be string overloads as well?

Reply:
Will add if there's use of it (possibly after looking at the string-related simplifications prompted by @WalterBright). Until then, ["Do the simplest thing..."](http://wiki.c2.com/?DoTheSimplestThingThatCouldPossiblyWork).